File: 559.Maximum-Depth-of-N-ary-Tree.py
# https://leetcode.com/problems/maximum-depth-of-n-ary-tree/description/
#
# algorithms
# Easy (41.9%)
# Total Accepted:    10k
# Total Submissions: 16.4k
# beats 61.0% of python submissions


"""
# Definition for a Node.
class Node(object):
    def __init__(self, val, children):
        self.val = val
        self.children = children
"""


class Solution(object):
    def maxDepth(self, root):
        """
        :type root: Node
        :rtype: int
        """
        if not root:
            return 0

        if not root.children:
            return 1

        max_child_depth = float('-inf')

        for child in root.children:
            depth = self.maxDepth(child)
            if depth > max_child_depth:
                max_child_depth = depth

        return max_child_depth + 1

File: 559.Maximum-Depth-of-N-ary-Tree.java
// https://leetcode.com/problems/maximum-depth-of-n-ary-tree/
//
// algorithms
// Easy (66.35%)
// Total Accepted: 73,257
// Total Submissions: 110,418
// beats 100.0% of java submissions


/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val,List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/
class Solution {

    static int res;

    public int maxDepth(Node root) {
        if (root == null) {
            return 0;
        }
        res = 1;

        recursive(root, 1);

        return res;
    }

    public void recursive(Node node, int depth) {
        if (node.children == null || node.children.size() == 0) {
            res = Math.max(res, depth);
            return;
        }

        for (Node c : node.children) {
            recursive(c, depth + 1);
        }
    }
}